proto/prom/find.go

```diff
@@ -104,40 +105,87 @@ func (p *QL) FindLabelsValues(ctx echo.Context) error {
 
 	token := core.RetrieveToken(r)
 	if len(token) == 0 {
-		respondWithError(w, errors.New("Not authorized, please provide a READ token"), http.StatusForbidden)
+		respondWithError(w, errors.New("not authorized, please provide a READ token"), http.StatusForbidden)
 		return nil
 	}
 
 	labelValue := ctx.Param("label")
-
 	if len(labelValue) == 0 {
-		log.WithFields(log.Fields{}).Error("missing label")
-		respondWithError(w, errors.New("Unprocessable Entity: label"), http.StatusBadRequest)
+		log.Error("missing label")
+		respondWithError(w, errors.New("unprocessable Entity: label"), http.StatusBadRequest)
+		return nil
+	}
+
+	// Parse form data
+	if err := r.ParseForm(); err != nil {
+		respondWithError(w, errors.New("failed to parse form data"), http.StatusBadRequest)
+		return nil
+	}
+
+	// Get matchers and process them
+	matchers := r.Form["match[]"]
+	if len(matchers) == 0 {
+		// Grafana will try to get all class name when arriving explore page
+		// This prevent showing an error to the customer, while allowing to prevent performance
+		// bottleneck where the user is fetching 1M series
+		resp := prometheusFindLabelsResponse{
+			Status: "success",
+			Data:   []string{},
+		}
+		return ctx.JSON(http.StatusOK, resp)
+	}
+
+	// TODO: Prevent issues by parsing the start param
+
+	// Process the first matcher to build the Warp10 selector
+	matcherObjs, err := queryPromql.ParseMetricSelector(matchers[0])
+	if err != nil {
+		respondWithError(w, err, http.StatusBadRequest)
 		return nil
 	}
 
-	selector := "~.*{" + labelValue + "~.*}"
+	// Extract class name and build labels map
+	classname := ""
+	labels := make(map[string]string)
+
+	for _, matcher := range matcherObjs {
```

Comment:
you can extract it in a function and add some tests

Reply:
Done with a test suite